### PR TITLE
Refactoring plus Documentation

### LINKED ISSUE #2: Fix xgb converter for version > 0.90
Regression for XGBoost is failing for 1.0.2. 

Apparently there was:
- a change in the API, ``alpha`` is now an array\list
- a change in the model structure because we are getting different results for regression
- a change in ``max_depth`` because ``None`` does not work as ``max_depth`` in 0.90

### LINKED ISSUE #4: params to determine tree type, and max_depth
right now, we are hardcoding 3 or 4 as low, and 10 as high. We need to do more testing and tuning and document our choices.

Also, there are a lot of inconsistencies and duplications with the `max_depth` parameter. We need to revisit that for all 4 trees.

### LINKED ISSUE #7: RF beam++ single node
There is a bug in beam++ for RF with node size 1 somewhere related to tree_commons.py:481

### PR DESCRIPTION
**Pull request by interesaaat**
This PR does some major refactoring. Less code and more documentation. Will need to do another pass to finish the documentation but we are getting there. Please don't hate me.

Close #7.
Close #2. 
Close #4 .

### CODE REVIEW

File: hummingbird/__init__.py [hunk 1]
```diff
@@ -5,7 +5,7 @@
 # --------------------------------------------------------------------------
 
 """
-Main entry point to the converter from the *scikit-learn* to *PyTorch*.
+Main entry point to the Humminbird converters.
```
Comment (ksaur) [inline hummingbird/__init__.py]: nit: missing 'g'

File: hummingbird/_container.py [hunk 2]
```diff
@@ -46,9 +46,9 @@ def output_names(self):
 
 class SklearnModelContainerNode(RawModelContainerNode):
     """
-    Main container for one *scikit-learn* model.
+    Main container for a *scikit-learn* model.
     Every converter adds nodes to an existing container
-    which is converted into a *PyTorch* graph by an instance of
+    which is the converted into a tensor graph by an instance of
```
Comment (ksaur) [inline hummingbird/_container.py]: s/the/then/g

File: hummingbird/_supported_operators.py [hunk 3]
```diff
@@ -39,20 +38,20 @@ def build_sklearn_operator_name_map():
     return res
 
 
-def _get_sklearn_operator_name(model_type):
+def get_sklearn_api_operator_name(model_type):
     """
-    Get operator name of the input argument
+    Get operator name of the input model in *scikit-learn API* format.
 
-    :param model_type:  A scikit-learn object (e.g., SGDClassifier
```
Comment (ksaur) [inline hummingbird/_supported_operators.py:46]: ooh good catch! :)

File: hummingbird/utils.py [hunk 4]
```diff
@@ -0,0 +1,58 @@
+# -------------------------------------------------------------------------
```
Comment (ksaur) [inline hummingbird/utils.py:1]: This is a nice change! I'm curious, did you write this file or is it from somewhere else that  I didn't see before?

File: hummingbird/_topology.py [hunk 5]
```diff
@@ -121,40 +127,36 @@ def original_operator(self):
 
 class Topology:
     """
-    Holds instances on :class:`Scope <hummingbird.common._topology.Scope>` and
-    :class:`SklearnModelContainer <hummingbird.common._container.SklearnModelContainer>`.
+    Holds instances on :class:`SklearnModelContainer <hummingbird.common._container.SklearnModelContainer>`.
```
Comment (ksaur) [inline hummingbird/_topology.py]: hummingbird/common/_container.py → hummingbird/_container.py

File: hummingbird/_topology.py [hunk 6]
```diff
@@ -199,9 +201,6 @@ def _generate_unique_name(seed, existing_names):
             existing_names.add(new_name)
             return new_name
 
-    def get_unique_scope_name(self, seed):
```
Comment (ksaur) [inline hummingbird/_topology.py]: How was this code ever used before? At one point was there some additional scope functionality?

File: hummingbird/_topology.py [hunk 7]
```diff
@@ -262,8 +269,8 @@ def topological_operator_iterator(self):
         This is an iterator of all operators in Topology object.
         Operators may be produced in a topological order. If you want to
         simply go though all operators without considering their
-        topological structure, please use another function,
-        unordered_operator_iterator.
+        topological structure, please use
+        :meth:`Topology.compile <hummingbird.common._topology.Topology.unordered_operator_iterator>`.
```
Comment (ksaur) [inline hummingbird/_topology.py]: hummingbird/common/_topology.py → hummingbird/_topology.py

File: hummingbird/_topology.py [hunk 8]
```diff
@@ -349,7 +355,10 @@ def convert_topology(topology, device=None, extra_config={}):
...
             )
 
-    pytorch_model = HBPyTorchModel(
-        topology.raw_model.input_names, topology.raw_model.output_names, operator_map, topology, device, extra_config
-    )
+    pytorch_model = PyTorchBackendModel(
+        topology.raw_model.input_names, topology.raw_model.output_names, operator_map, topology, extra_config
+    ).eval()
+
+    if device is not None:
```
Comment (ksaur) [inline hummingbird/_topology.py]: nice work on moving this here, much cleaner!

File: hummingbird/convert.py [hunk 9]
```diff
@@ -6,78 +6,71 @@
...
-"""
-This function is used to generate a PyTorch model from a given LighGBM model
-:param model: A LightGBM model (trained using the scikit-learn API)
-:param test_input: some input data used to trace the model execution
-:param device: torch.device Which device to translate them model
-:param extra_config: Extra configurations to be used by the individual operator converters
-:return: A PyTorch model which is equivalent to the input LightGBM model
-"""
+    # Convert our Topology object into PyTorch.
+    hb_model = convert_topology(topology, device, extra_config)
```
Comment (ksaur) [inline hummingbird/convert.py]: any reason to not just return directly and skip a line of code?

File: hummingbird/convert.py [hunk 10]
```diff
@@ -6,78 +6,71 @@
...
+    return hb_model
 
 
 def convert_lightgbm(model, test_input=None, device=None, extra_config={}):
-    return convert_sklearn(model, test_input, device, extra_config)
-
+    """
+    This function is used to generate a PyTorch model from a given LighGBM model
```
Comment (ksaur) [inline hummingbird/convert.py]: s/LighGBM/LightGBM/g

File: hummingbird/convert.py [hunk 11]
```diff
@@ -6,78 +6,71 @@
...
+    :param model: A LightGBM model (trained using the scikit-learn API)
+    :param test_input: some input data used to trace the model execution
+    :param backend: Which backend to translate the model to. PyTorch by default
```
Comment (ksaur) [inline hummingbird/convert.py]: Is `Which backend to translate..` confusing to the user for now? Since we only talk about pyt?

File: hummingbird/_topology.py [hunk 12]
```diff
@@ -121,40 +127,36 @@ def original_operator(self):
...
 class Topology:
     """
-    Holds instances on :class:`Scope <hummingbird.common._topology.Scope>` and
-    :class:`SklearnModelContainer <hummingbird.common._container.SklearnModelContainer>`.
+    Holds instances on :class:`SklearnModelContainer <hummingbird.common._container.SklearnModelContainer>`.
     These are filled by the converters while a pipeline is being converted.
-    When all converters were called, method
-    :meth:`Topology.compile <hummingbird.common._topology.Topology.compile>`
-    must be called to convert the topological graph into PyTorch model.
+    When all converters were called, method :meth:`Topology.compile <hummingbird.common._topology.Topology.compile>`
```
Comment (ksaur) [inline hummingbird/_topology.py]: hummingbird/common/_topology.py → hummingbird/_topology.py

File: hummingbird/convert.py [hunk 13]
```diff
@@ -6,78 +6,71 @@
...
 from . import operator_converters  # noqa
 
 
 def convert_sklearn(model, test_input=None, device=None, extra_config={}):
     """
-    This function produces an equivalent PyTorch model of the given scikit-learn model.
-    The supported converters is returned by function
-    :func:`supported_converters <hummingbird.supported_converters>`.
+    This function produces, given a scikit-learn model, an equivalent model in the selected backend.
+    The supported operators and backends be found at :func:`supported_converters <hummingbird._supported_opeartors>`.
```
Comment (ksaur) [inline hummingbird/convert.py]: _supported_opeartors   Either fix spelling on  operators, or..should this be `converters` instead?

File: hummingbird/convert.py [hunk 14]
```diff
@@ -6,78 +6,71 @@
...
     is one of our supported items.
 
     This function converts the specified *scikit-learn* model into its *PyTorch* counterpart.
-    Note that for all conversions, initial types are required.
-    *TorchScript* model file name can also be specified.
 
     :param model: A scikit-learn model
     :param test_input: some input data used to trace the model execution
-    :param device: torch.device Which device to translate them model
+    :param device: Which device to translate the model to. CPU by defauls
```
Comment (ksaur) [inline hummingbird/convert.py]: s/defauls/default/g    (this occurs  in 3 places)

File: hummingbird/operator_converters/_gbdt_commons.py [hunk 15]
```diff
@@ -0,0 +1,173 @@
...
+# --------------------------------------------------------------------------
+
+import torch
+import numpy as np
+
+from ._tree_commons import get_tree_params_and_type, get_parameters_for_tree_trav_common, get_parameters_for_gemm_common
+from ._tree_commons import GEMMTreeEnsemble, TreeTraversalTreeEnsemble, PerfectTreeTraversalTreeEnsemble, TreeImpl
+
+
+class GEMMGBDT(GEMMTreeEnsemble):
```
Comment (ksaur) [inline hummingbird/operator_converters/_gbdt_commons.py]: Is this just a "lift and shift" from the other file, or should I be looking more closely here?


BTW, nice, change, this is a much better organization.

File: hummingbird/operator_converters/_gbdt_commons.py [hunk 16]
```diff
@@ -0,0 +1,173 @@
...
+
+    def calibation(self, x):
+        if self.binary_classification:
+            output = torch.sigmoid(x)
+            return torch.cat([1 - output, output], dim=1)
+        else:
+            return torch.softmax(x, dim=1)
+
+
+def convert_gbdt_classifier_common(
```
Comment (ksaur) [inline hummingbird/operator_converters/_gbdt_commons.py:112]: I think autopep got too aggressive here. Maybe change this back to look like  `convert_gbdt_common` params?

File: hummingbird/operator_converters/_gbdt_commons.py [hunk 17]
```diff
@@ -0,0 +1,173 @@
...
+    if tree_type == TreeImpl.gemm:
+        net_parameters = [
+            get_parameters_for_gemm_common(
+                tree_param.lefts, tree_param.rights, tree_param.features, tree_param.thresholds, tree_param.values, n_features
+            )
+            for tree_param in tree_parameters
+        ]
+        return GEMMGBDT(net_parameters, n_features, classes, learning_rate, alpha, device)
+
+    # Some model requires some additional massagging of the parameters before generating the tree_trav implementation.
```
Comment (ksaur) [inline hummingbird/operator_converters/_gbdt_commons.py]: I think this should read as 
`Some models require some additional massaging of..`

File: hummingbird/operator_converters/_tree_commons.py [hunk 18]
```diff
@@ -20,211 +33,99 @@ def __init__(self, id):
...
-    values = np.array(values)
-
-    return [depth, nodes_map, ids, lefts, rights, features, thresholds, values]
-
-
-def get_parameters_for_tree_trav_sklearn_estimators(tree):
-    """This function is used by sklearn-based trees.
-
-    Includes SklearnRandomForestClassifier/Regressor and SklearnGradientBoostingClassifier
+class AbstractTreeEnsemble(ABC):
```
Comment (ksaur) [inline hummingbird/operator_converters/_tree_commons.py]: Ok, just so I understand, this  `AbstractTreeEnsemble ` is now the "base class", and then only `AbstractPyTorchTreeEnsemble` adds a bit more? And then from there, `AbstractPyTorchTreeEnsemble` leads to all (current) others.  Is your plan for `AbstractTreeEnsemble` to be used for future backends?

File: hummingbird/operator_converters/_tree_commons.py [hunk 19]
```diff
@@ -296,11 +220,11 @@ def __init__(self, tree_parameters, n_features, n_classes):
         values = np.zeros((self.num_trees, self.num_nodes, self.n_classes), dtype=np.float32)
 
         for i in range(self.num_trees):
-            lefts[i][: len(tree_parameters[i][1])] = tree_parameters[i][3]
-            rights[i][: len(tree_parameters[i][1])] = tree_parameters[i][4]
-            features[i][: len(tree_parameters[i][1])] = tree_parameters[i][5]
-            thresholds[i][: len(tree_parameters[i][1])] = tree_parameters[i][6]
-            values[i][: len(tree_parameters[i][1])][:] = tree_parameters[i][7]
+            lefts[i][: len(tree_parameters[i][0])] = tree_parameters[i][2]
```
Comment (ksaur) [inline hummingbird/operator_converters/_tree_commons.py]: I'm curious why this change?

File: hummingbird/operator_converters/_tree_commons.py (hunk 18 above)
Comment (ksaur) [reply hummingbird/operator_converters/_tree_commons.py]: and `AbstractPyTorchTreeEnsemble` is just random forest?  Did you state this ('ensemble' ~= 'random forest') somewhere in case someone is more familiar with that term?

File: hummingbird/operator_converters/_tree_commons.py [hunk 20]
```diff
@@ -462,3 +429,268 @@ def depth_f_traversal(node, current_depth, node_id, leaf_start_id):
...
+    Populate the parameters from the trees and pick the tree implementation strategy.
+    """
+    tree_parameters = [get_tree_parameters(tree_info) for tree_info in tree_infos]
+    max_depth = max(1, _find_max_depth(tree_parameters))
+    tree_type = get_tree_implementation_by_config_or_depth(extra_config, max_depth)
+
+    return tree_parameters, max_depth, tree_type
+
+
+def get_parameters_sklearn_common(tree_info):
```
Comment (ksaur) [inline hummingbird/operator_converters/_tree_commons.py]: many of the other functions are named `get_blah_FOR_foo_common`...this seems like it's missing the `for`.  I don't feel strongly; just an observation

File: hummingbird/operator_converters/_tree_commons.py [hunk 21]
```diff
@@ -20,211 +33,99 @@ def __init__(self, id):
...
+        self.lefts = lefts
+        self.rights = rights
+        self.features = features
+        self.thresholds = thresholds
+        self.values = values
 
 
 class TreeImpl(Enum):
+    """
+    Enum defininig the available implementations for tree scoring.
```
Comment (ksaur) [inline hummingbird/operator_converters/_tree_commons.py]: s/defininig/defining/g

File: hummingbird/operator_converters/_gbdt_commons.py [hunk 22]
```diff
@@ -0,0 +1,173 @@
+# -------------------------------------------------------------------------
```
Comment (ksaur) [inline hummingbird/operator_converters/_gbdt_commons.py:1]: It's  really confusing to me how we have a file called `tree_commons` and also `gbdt_commons`. Can you add some text at the top of each of the files explaining?

File: hummingbird/operator_converters/decision_tree.py [hunk 23]
```diff
@@ -0,0 +1,156 @@
+# -------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See License.txt in the project root for
+# license information.
+# --------------------------------------------------------------------------
+import warnings
+import copy
+
+import torch
+
```
Comment (ksaur) [inline hummingbird/operator_converters/decision_tree.py:11]: Ok  so  this is still `RandomForest` but it's also....something else too now? Can  you please add some text explaining?   It's confusing to me (and maybe others) how this  does/doesn't map  to scikit learn

File: hummingbird/operator_converters/_tree_commons.py [hunk 24]
```diff
@@ -462,3 +429,268 @@ def depth_f_traversal(node, current_depth, node_id, leaf_start_id):
...
+            path.pop()
+
+    weights.append(np.array(hidden_weights).astype("float32"))
+    biases.append(np.array(hidden_biases).astype("float32"))
+
+    # OR neurons from the preceding layer in order to get final classes.
+    weights.append(np.transpose(np.array(class_proba).astype("float32")))
+    biases.append(None)
+
+    return weights, biases
```
Comment (ksaur) [inline hummingbird/operator_converters/_tree_commons.py:322]: Heads-up that this file is now nearly 700 lines...It think there are trade-offs to having all in one giant file vs some smaller separate  ones. it's ok  but if it grows again we might need to re-org

**Review by ksaur — COMMENTED** [review]
Overall, this looks great! 
I just had a few minor comments and several questions.  

I noticed that coverage dropped from 92% to 90%…I dug around for a while and then I determined that we actually just have less code to artificially inflate the percentage, so I think we’re good there. :)

File: hummingbird/__init__.py (hunk 1 above)
Comment (interesaaat) [reply hummingbird/__init__.py]: damn it! Hummingbird is really hard to spell :)

File: hummingbird/utils.py (hunk 4 above)
Comment (interesaaat) [reply hummingbird/utils.py:1]: From onnxmltools :) I saw this while implementing the API and thought that is actually I nice thing to have.

File: hummingbird/_topology.py (hunk 6 above)
Comment (interesaaat) [reply hummingbird/_topology.py]: I was able to remove topology. So no more doubts around this part of the code.

File: hummingbird/convert.py (hunk 10 above)
Comment (interesaaat) [reply hummingbird/convert.py]: good catch!

File: hummingbird/convert.py (hunk 11 above)
Comment (interesaaat) [reply hummingbird/convert.py]: Yep. This is a typo. I started to implement the backend part but then stopped. Will resume once we have other backends.

File: hummingbird/convert.py (hunk 13 above)
Comment (interesaaat) [reply hummingbird/convert.py]: Fixed to `_supported_operators`

File: hummingbird/operator_converters/_tree_commons.py (hunk 19 above)
Comment (interesaaat) [reply hummingbird/operator_converters/_tree_commons.py]: Ah I got rid of the `max_depth` parameter. Now it is passed as input to the `class __init__`, so I had to shift all indices

File: hummingbird/operator_converters/_gbdt_commons.py (hunk 16 above)
Comment (interesaaat) [reply hummingbird/operator_converters/_gbdt_commons.py:112]: I tried and failed because it is > 127 chars.

File: hummingbird/operator_converters/_gbdt_commons.py (hunk 15 above)
Comment (interesaaat) [reply hummingbird/operator_converters/_gbdt_commons.py]: Yea I basically moved everything in common in a single file and do a couple of iterations to cut the unnecessary.

File: hummingbird/operator_converters/_tree_commons.py (hunk 18 above)
Comment (interesaaat) [reply hummingbird/operator_converters/_tree_commons.py]: Yes I wanted to have a couple of abstract methods (`aggregation` and `calibration`). And yes, `AbstractTreeEnsemble` will be the base class for other backends too. Nope it is not "just" random forest but actually all trees (decisions trees, gbdt, etc.). Maybe I should change it into just `AbstractTree` or `AbstractDecisionTree`

**Comment by interesaaat** [comment]
> Overall, this looks great!
> I just had a few minor comments and several questions.
> 
> I noticed that coverage dropped from 92% to 90%…I dug around for a while and then I determined that we actually just have less code to artificially inflate the percentage, so I think we’re good there. :)

We are at 94% now. I got rid of _topology :)

File: hummingbird/operator_converters/_tree_commons.py (hunk 20 above)
Comment (interesaaat) [reply hummingbird/operator_converters/_tree_commons.py]: Good catch! Fixed.

File: hummingbird/operator_converters/_tree_commons.py (hunk 21 above)
Comment (interesaaat) [reply hummingbird/operator_converters/_tree_commons.py]: Good catch! Fixed.

File: hummingbird/operator_converters/_gbdt_commons.py (hunk 22 above)
Comment (interesaaat) [reply hummingbird/operator_converters/_gbdt_commons.py:1]: Yep. Tree commons is for all tree algorithms. GBDT commons is only for GBDT trees.

File: hummingbird/operator_converters/decision_tree.py (hunk 23 above)
Comment (interesaaat) [reply hummingbird/operator_converters/decision_tree.py:11]: Yea. I think DecisionTree is more generic than random forest. Like a random forest is an ensemble of decision trees so I thought that the main (base) file should be decision tree. Will add a comment.

File: hummingbird/operator_converters/_gbdt_commons.py (hunk 16 above)
Comment (ksaur) [reply hummingbird/operator_converters/_gbdt_commons.py:112]: Is something like:
```
def convert_gbdt_classifier_common(self, tree_infos, get_tree_parameters, n_features, n_classes, classes=None, \
             learning_rate=None, alpha=None, device=None):
```
with a slash at the end allowed?

File: hummingbird/operator_converters/_tree_commons.py (hunk 24 above)
Comment (interesaaat) [reply hummingbird/operator_converters/_tree_commons.py:322]: Good point! Let me split this in 2 files: `_tree_implementations` and `_tree_commons`.

**Comment by interesaaat** [comment]
Done with this round or comments!

File: hummingbird/operator_converters/_gbdt_commons.py (hunk 16 above)
Comment (interesaaat) [reply hummingbird/operator_converters/_gbdt_commons.py:112]: Lemme try.

Comment (interesaaat) [reply hummingbird/operator_converters/_gbdt_commons.py:112]: Nope it is not.

File: hummingbird/operator_converters/decision_tree.py (hunk 23 above)
Comment (ksaur) [reply hummingbird/operator_converters/decision_tree.py:11]: Oh, yah, it was not obvious (at least to me) that `random forest is an ensemble of decision trees` so that would be helpful!
